File: docs/truth_assurance_pipeline/experiments/tap_e7_artifact_assurance/tap-e7-base-implementation-b/tools/run_blind_conformance.js
'use strict';
// Blind harness: wraps fs.readFileSync to log every package read and THROW on any
// expected/ or derivations/ read while blind is active. Projects each fixture to the
// true input envelope only.
const fs=require('fs'), path=require('path');
const V=require('../src/verifier.js');
const CF=require('./recompute_config_fingerprint.js');
const PKG=process.argv[2];
const OUT=path.join(__dirname,'..','results');
const realRead=fs.readFileSync;
const OPENED=[]; let BLIND=false;
fs.readFileSync=function(f,...a){ const p=String(f);
  if(p.includes('tap-e7-base-companion-1.1.1')) OPENED.push(p);
  if(BLIND && (p.includes('/expected/')||p.includes('/derivations/')||p.includes('-audit/')||p.includes('implementation-a')))
    throw new Error('BLIND VIOLATION: '+p);
  return realRead.call(fs,f,...a); };

const [got,want]=CF.recompute(PKG);
if(got!==want){ console.error('FINGERPRINT MISMATCH — stop'); process.exit(2); }
const res=V.loadResources(PKG);
const cm=JSON.parse(realRead(path.join(PKG,'manifest/corpus-manifest.json')));
const authoritative={}; for(const e of cm.fixtures){ authoritative[path.basename(e.path).slice(0,-5)]=e.authoritative; }
const INPUT=['modality','validation_record','artifact','profile_ref','release_ref'];
fs.mkdirSync(path.join(OUT,'blind'),{recursive:true});
const corpusDir=path.join(PKG,'corpus'); const produced={}; const timings={};
for(const fn of fs.readdirSync(corpusDir).sort()){
  const fid=fn.slice(0,-5);
  const obj=JSON.parse(realRead(path.join(corpusDir,fn)));
  const sub={}; for(const k of INPUT) if(k in obj) sub[k]=obj[k];
  BLIND=true; const t0=process.hrtime.bigint();
  const rec=V.evaluate(res,sub);
  const tr=V.trace(sub,rec,false), trr=V.trace(sub,rec,true);
  const dt=Number(process.hrtime.bigint()-t0)/1e6; BLIND=false;
  timings[fid]=dt;
  const out={fixture_id:fid,authoritative:authoritative[fid],assurance_record:rec,trace:tr,redacted_trace:trr};
  fs.writeFileSync(path.join(OUT,'blind',fid+'.json'),JSON.stringify(out,null,1));
  produced[fid]=out;
}
const expReads=OPENED.filter(p=>p.includes('/expected/')||p.includes('/derivations/'));
const proof={fixtures_evaluated:Object.keys(produced).length,expected_or_derivation_reads_during_blind:expReads,
  blind_boundary_intact:expReads.length===0,config_fingerprint_recomputed:got,config_fingerprint_match:got===want};
fs.writeFileSync(path.join(__dirname,'..','clean_room_evidence','blind-boundary-proof.json'),JSON.stringify(proof,null,1));
// access log: unique package paths read
const uniq=[...new Set(OPENED)].map(p=>p.split('tap-e7-base-companion-1.1.1/')[1]||p).sort();
fs.writeFileSync(path.join(__dirname,'..','clean_room_evidence','file-access-log.json'),
  JSON.stringify({package_paths_read:uniq,any_forbidden:uniq.some(p=>p.startsWith('expected/')||p.startsWith('derivations/'))},null,1));
const crypto=require('crypto'); const h=crypto.createHash('sha256');
for(const fn of fs.readdirSync(path.join(OUT,'blind')).sort()) h.update(realRead(path.join(OUT,'blind',fn)));
fs.writeFileSync(path.join(__dirname,'..','clean_room_evidence','blind-output-root.json'),
  JSON.stringify({blind_output_sha256:'sha-256:'+h.digest('hex'),fixtures:Object.keys(produced).length},null,1));
fs.writeFileSync(path.join(OUT,'timings.json'),JSON.stringify(timings,null,1));
console.log('blind: evaluated',Object.keys(produced).length,'| boundary intact:',proof.blind_boundary_intact,'| fingerprint:',got===want);

File: docs/truth_assurance_pipeline/experiments/tap_e7_artifact_assurance/tap-e7-base-implementation-a/tools/run_blind_conformance.py
#!/usr/bin/env python3
"""Blind conformance harness (§24).
Phase 1: expose ONLY the CandidateArtifact + ValidationRecord + descriptor envelope to
the verifier; hide expected/, derivations/, and every expected/metadata field embedded in
the corpus fixture. Enforce the blind boundary with an audit-open wrapper that records every
package path opened and hard-fails if expected/ or derivations/ is read during evaluation."""
import builtins, hashlib, json, os, sys, time
HERE = os.path.dirname(os.path.abspath(__file__))
SRC = os.path.join(HERE, "..", "src"); sys.path.insert(0, SRC)
import verifier as V

PKG = sys.argv[1]
OUT = os.path.join(HERE, "..", "results")
os.makedirs(os.path.join(OUT, "produced"), exist_ok=True)

# ---- config fingerprint gate ----
sys.path.insert(0, HERE)
import recompute_config_fingerprint as CF
got, want = CF.recompute(PKG)
if got != want:
    print("CONFIG FINGERPRINT MISMATCH — stopping", got, want); sys.exit(2)

# ---- blind-read boundary: forbid expected/ and derivations/ during evaluation ----
_open = builtins.open
OPENED = []
BLIND = {"active": False}
def guarded_open(file, *a, **k):
    p = str(file)
    if PKG in p or p.startswith(("expected/", "derivations/")):
        OPENED.append(p)
        if BLIND["active"] and (os.sep + "expected" + os.sep in p or os.sep + "derivations" + os.sep in p):
            raise PermissionError("BLIND VIOLATION: verifier attempted to read " + p)
    return _open(file, *a, **k)
builtins.open = guarded_open

ver = V.Verifier(PKG)                                   # loads resources (allowed)
corpus_dir = os.path.join(PKG, "corpus")
manifest = json.load(_open(os.path.join(PKG, "manifest/corpus-manifest.json")))
authoritative = {}
for e in manifest["fixtures"]:
    fid = os.path.basename(e["path"])[:-5]
    authoritative[fid] = e.get("authoritative", True)

INPUT_FIELDS = ("modality", "validation_record", "artifact", "profile_ref", "release_ref")
def blind_submission(fixture_obj):
    """Strip everything except the true verifier inputs (drop expected/phenomenon/purpose/etc.)."""
    return {k: fixture_obj[k] for k in INPUT_FIELDS if k in fixture_obj}

produced = {}
timings = {}
for fn in sorted(os.listdir(corpus_dir)):
    fid = fn[:-5]
    obj = json.load(_open(os.path.join(corpus_dir, fn)))
    sub = blind_submission(obj)
    BLIND["active"] = True
    t0 = time.perf_counter()
    rec = ver.evaluate(sub)
    trace_full = ver.trace(sub, rec, redacted=False)
    trace_red = ver.trace(sub, rec, redacted=True)
    dt = time.perf_counter() - t0
    BLIND["active"] = False
    timings[fid] = dt
    out = {"fixture_id": fid, "authoritative": authoritative.get(fid, True),
           "assurance_record": rec, "trace": trace_full, "redacted_trace": trace_red}
    json.dump(out, _open(os.path.join(OUT, "produced", fid + ".json"), "w"), sort_keys=True, indent=1)
    produced[fid] = out

builtins.open = _open
# proof of blind boundary
expected_reads = [p for p in OPENED if (os.sep + "expected" + os.sep in p or os.sep + "derivations" + os.sep in p)]
proof = {"fixtures_evaluated": len(produced),
         "expected_or_derivation_reads_during_blind": expected_reads,
         "blind_boundary_intact": len(expected_reads) == 0,
         "config_fingerprint_recomputed": got, "config_fingerprint_match": got == want}
json.dump(proof, _open(os.path.join(OUT, "blind-proof.json"), "w"), indent=1)
json.dump(timings, _open(os.path.join(OUT, "timings.json"), "w"), indent=1)
print("phase1 blind: evaluated", len(produced), "| blind boundary intact:", proof["blind_boundary_intact"],
      "| fingerprint match:", got == want)
